rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+
+4. Bells And Whistles: Local rollbacks, pending/approval flow, client upgrades, etc...
+
+
+## Details
+
+### Usage Scenarios
+
+TODO: update these sections with improved terminology.
+
+#### Minimal Usecase
+
+Cluster administrators explicitly set the desired version(s) of teleport installations, and
+have a specific in-house installation script that should be used for upgrades. Teleport cluster
+may not even have any access to the public internet.  The installation target may be an internally
+built fork of teleport.
+
+In this case, we want to provide the means to specify the target version and desired script. Teleport
+should then be able to detect when a teleport node is not running the target that it aught to, and
+invoke the upgrade script.  Teleport should not care about the internals of how the script plans to
+perform the installation.  Nodes that require upgrades run the script, and may be required to perform
+a graceful restart if the script succeeds. The script may expect inputs (e.g. version string), and
+there may be different scripts to run depending on the nature of the specific node (e.g. `prod-upgrade`, `staging-upgrade`
+or `apt-upgrade`, `yum-upgrade`), but things like selecting the correct processor architecture or OS
+compatibility are likely handled by the script.
+
+In this minimal usecase teleport's role is primarily as a coordinator.  It detects when and were user
+provided scipts should be run, and invokes them.
+
+#### Maximal Usecase
+
+Cluster administrators opt into a mostly "set and forget" upgrade policy which keeps their teleport
+cluster up to date automatically. They may wish to stay at a specific major version, but would like
+to have patches and minor backwards-compatible improvements come in automatically.  They want features
+like maintenance schedules to prevent a node from being upgraded when they need it available, and automatic
+rollbacks where nodes revert to their previous installation version of they are unhealthy for too long.
+They also want teleport to be able to upgrade itself without dependencies.
+
+This usecase requires the same coordination powers as the minimal usecase, but also a lot more.  Teleport
+needs to be able to securely and reliably detect new releases when they become availalbe.  Teleport needs
+to be able to evaluate new releases in the context of flexible upgrade policies and select which releases
+(and which targets within those releases) are appropriate and when they should be installed. Teleport needs
+to be able to download and verify installation packages, upgrade itself, and monitor the health of newly
+installed versions.
+
+In this maximal usecase, teleport is responsible for discovery, selection, coordination, validation, and
+monitoring.  Most importantly, teleport must do all of this in a secure and reliable manner.  The potential
+fallout from bugs and vulnerabilities is greater than in the minimal usecase.
+
+#### Hacker Usecase
```

Comment:
Can we have two other use-cases? 

* Dry run use-case is always helpful so it can tell you what will happen without actually running things.

Dry run use-case is also helpful to mitigate massive network spike. In dry-run use-case packages can be fetched locally on each machine, but not actually installed. This will help to avoid another thundering herd issue with 10K teleport nodes pulling the packages and restarting.

* Alert but don't upgrade use-case can set some internal "alert" state if your system is out of date. It is similar to dry run, but this can be turned on by default. Kind of like Chrome "green/yellow/red" button

To me, the biggest value of this system even before the upgrade would be to have a periodic job on teleport side that checks against TUF, does not upgrade anything, but sets internal alert in the dashboard or teleport status that tells that system is running vulnerable software. I'd like to include this in MVP - this will solve a lot of issues for many users who are not aware of the vulnerabilities today.